### ISSUE TITLE
Guava version conflicts - NoSuchMethodError

### ISSUE THREAD

**Issue by Wilwilly** [ronmamo/reflections#194]
I'm getting the following exception using Guava v23 and Reflections v0.9.11.

```
Exception in thread "main" java.lang.NoSuchMethodError: com.google.common.collect.Sets$SetView.iterator()Lcom/google/common/collect/UnmodifiableIterator;
	at org.reflections.Reflections.expandSuperTypes(Reflections.java:380)
	at org.reflections.Reflections.<init>(Reflections.java:126)
	at org.reflections.Reflections.<init>(Reflections.java:168)
	at org.reflections.Reflections.<init>(Reflections.java:141)
        ...
```

Going back to Reflections v0.9.10 fixes the problem, since it was compiled for Guava v15, which seems to be somehow compatible with Guava v23 but not with v20 (version used in Reflections v0.9.11).

Any chance to update this lib to use Guava v23 in future releases? That would fix the problem once and for all.

**Comment by stanjg** [comment]
I have this issue as well, is there any fix for it?

**Comment by scandinave** [comment]
Same here

**Comment by Flexximilian** [comment]
Same here, and everywhere, really. Just google the error. Guava is evil. 

They change things in a binary incompatible way without any regard to the trouble they cause, or the cost they create. Now that's fine for internal packages, but not for something they put out without warning that such is their governance model. Newer versions break existing code built against older versions, all in the same classes and packages (unlike Apache commons, who make very sure that code doesn't break, and if it does it goes in a separate package. Hence lang3 and collections4, etc.). It's very easy to end up with a maven dependency tree of software requiring guava versions containing methods that are mutually existent/non-existent in different versions. So not even a rebuild can fix the mess. There is no chance of fixing this if two such pieces of code are 3rd party libraries depending on guava. If you're lucky, one of the libraries is open source with a permissive licence where a fork is an option.

I'd vote for reflections to ditch the dependency on guava altogether.

**Comment by obykoo** [comment]
Same problem. Any fix without going back to v0,9,10?

**Comment by Rentier** [comment]
Maybe we could look into using http://imperceptiblethoughts.com/shadow/ . It looks like it isolates the Guava reflections uses so other applications depending on it do not pull it and vice versa.

**Comment by jshook** [comment]
Any chance we could simply remove the guava dependency and use other options from Java 8, for example? The broken version dependencies in guava have cause significant pain in downstream projects, requiring shading and other techniques. It is difficult to need the functionality of this library but to have to micromanage around guava dependencies from multiple projects at the same time.

I've started excising guava from all my top-level projects in order to simplify builds back down. It would be really great if I didn't have to include it for reflections too.

**Comment by UnopenedClosure** [comment]
I found a workaround which might be helpful to those who can affect the creation of the Reflections object (my project uses reflections version 0.9.1 and guava version 16.0.1):

1. Create a class VersionCompliantReflections which extends Reflections
2. Create dummy implementations of the constructors (in other words, just call the super() for the same set of parameters)
3. "Copy" a couple of method implementations out of Reflections into VersionCompliantReflections:

	    public static String index(Class<? extends Scanner> scannerClass) {
	        return scannerClass.getSimpleName();
	    }
		
		@Override
	    public void expandSuperTypes() {
	        if (store.keySet().contains(index(SubTypesScanner.class))) {
	            Multimap<String, String> mmap = store.get(index(SubTypesScanner.class));
	            Sets.SetView<String> keys = Sets.difference(mmap.keySet(), Sets.newHashSet(mmap.values()));
	            Multimap<String, String> expand = HashMultimap.create();
	            for (String key : keys) {
	                final Class<?> type = forName(key);
	                if (type != null) {
	                    expandSupertypes(expand, key, type);
	                }
	            }
	            mmap.putAll(expand);
	        }
	    }

	    private void expandSupertypes(Multimap<String, String> mmap, String key, Class<?> type) {
	        for (Class<?> supertype : ReflectionUtils.getSuperTypes(type)) {
	            if (mmap.put(supertype.getName(), key)) {
	                if (log != null) log.debug("expanded subtype {} -> {}", supertype.getName(), key);
	                expandSupertypes(mmap, supertype.getName(), supertype);
	            }
	        }
	    }
4. Instead of creating a Reflections object, create a VersionCompliantReflections object with the same parameters and method signatures.

**Comment by TinusJ** [comment]
Any update on this ?

**Comment by TinusJ** [comment]
> I found a workaround which might be helpful to those who can affect the creation of the Reflections object (my project uses reflections version 0.9.1 and guava version 16.0.1):
> 
> 1. Create a class VersionCompliantReflections which extends Reflections
> 2. Create dummy implementations of the constructors (in other words, just call the super() for the same set of parameters)
> 3. "Copy" a couple of method implementations out of Reflections into VersionCompliantReflections:
>    ```
>     public static String index(Class<? extends Scanner> scannerClass) {
>         return scannerClass.getSimpleName();
>     }
>     
>     @Override
>     public void expandSuperTypes() {
>         if (store.keySet().contains(index(SubTypesScanner.class))) {
>             Multimap<String, String> mmap = store.get(index(SubTypesScanner.class));
>             Sets.SetView<String> keys = Sets.difference(mmap.keySet(), Sets.newHashSet(mmap.values()));
>             Multimap<String, String> expand = HashMultimap.create();
>             for (String key : keys) {
>                 final Class<?> type = forName(key);
>                 if (type != null) {
>                     expandSupertypes(expand, key, type);
>                 }
>             }
>             mmap.putAll(expand);
>         }
>     }
>    
>     private void expandSupertypes(Multimap<String, String> mmap, String key, Class<?> type) {
>         for (Class<?> supertype : ReflectionUtils.getSuperTypes(type)) {
>             if (mmap.put(supertype.getName(), key)) {
>                 if (log != null) log.debug("expanded subtype {} -> {}", supertype.getName(), key);
>                 expandSupertypes(mmap, supertype.getName(), supertype);
>             }
>         }
>     }
>    ```
> 4. Instead of creating a Reflections object, create a VersionCompliantReflections object with the same parameters and method signatures.

Can you show the whole class please

**Comment by UnopenedClosure** [comment]
> > I found a workaround which might be helpful to those who can affect the creation of the Reflections object (my project uses reflections version 0.9.1 and guava version 16.0.1):
> > 
> > 1. Create a class VersionCompliantReflections which extends Reflections
> > 2. Create dummy implementations of the constructors (in other words, just call the super() for the same set of parameters)
> > 3. "Copy" a couple of method implementations out of Reflections into VersionCompliantReflections:
> >    ```
> >     public static String index(Class<? extends Scanner> scannerClass) {
> >         return scannerClass.getSimpleName();
> >     }
> >     
> >     @Override
> >     public void expandSuperTypes() {
> >         if (store.keySet().contains(index(SubTypesScanner.class))) {
> >             Multimap<String, String> mmap = store.get(index(SubTypesScanner.class));
> >             Sets.SetView<String> keys = Sets.difference(mmap.keySet(), Sets.newHashSet(mmap.values()));
> >             Multimap<String, String> expand = HashMultimap.create();
> >             for (String key : keys) {
> >                 final Class<?> type = forName(key);
> >                 if (type != null) {
> >                     expandSupertypes(expand, key, type);
> >                 }
> >             }
> >             mmap.putAll(expand);
> >         }
> >     }
> >    
> >     private void expandSupertypes(Multimap<String, String> mmap, String key, Class<?> type) {
> >         for (Class<?> supertype : ReflectionUtils.getSuperTypes(type)) {
> >             if (mmap.put(supertype.getName(), key)) {
> >                 if (log != null) log.debug("expanded subtype {} -> {}", supertype.getName(), key);
> >                 expandSupertypes(mmap, supertype.getName(), supertype);
> >             }
> >         }
> >     }
> >    ```
> > 4. Instead of creating a Reflections object, create a VersionCompliantReflections object with the same parameters and method signatures.
> 
> Can you show the whole class please

I've made a few tweaks specific to my own project's needs since I originally wrote this reply, but here's my current class:

`private static class VersionCompliantReflections extends Reflections {

    private static final String SCANNER_CLASS_NAME = SubTypesScanner.class.getSimpleName();
    
    public VersionCompliantReflections(Configuration configuration) { super(configuration); }
    
    @Override
    public <T> Set<Class<? extends T>> getSubTypesOf(final Class<T> type) {
        Set<Class<? extends T>> toReturn = new LinkedHashSet<Class<? extends T>>();
        for(String s : getAllFromStore(type)) {
            try {
                toReturn.add((Class<? extends T>) Class.forName(s));
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
        }
        return toReturn;
    }
    
    private <T> Iterable<String> getAllFromStore(Class<T> type) {
        Multimap<String, String> mmap = store.get(SCANNER_CLASS_NAME);
        Set<String> result = new LinkedHashSet<String>();
        Predicate<String> inputsFilter = configuration.getInputsFilter();
        for(String key : mmap.keySet()) {
            for(String v : mmap.get(key))
                if(inputsFilter.apply(v))
                    result.add(v);
            if(inputsFilter.apply(key))
                result.add(key);
        }
        return result;
    }
    
    @Override
    public void expandSuperTypes() {
        if (store.keySet().contains(SCANNER_CLASS_NAME)) {
            Multimap<String, String> mmap = store.get(SCANNER_CLASS_NAME);
            Sets.SetView<String> keys = Sets.difference(mmap.keySet(), Sets.newHashSet(mmap.values()));
            Multimap<String, String> expand = HashMultimap.create();
            for (String key : keys) {
                final Class<?> type = forName(key);
                if (type != null) {
                    expandSupertypes(expand, key, type);
                }
            }
            mmap.putAll(expand);
        }
    }

    private void expandSupertypes(Multimap<String, String> mmap, String key, Class<?> type) {
        for (Class<?> supertype : ReflectionUtils.getSuperTypes(type)) {
            if (mmap.put(supertype.getName(), key)) {
                if (log != null) log.debug("expanded subtype {} -> {}", supertype.getName(), key);
                expandSupertypes(mmap, supertype.getName(), supertype);
            }
        }
    }        
}`

Usage at this point is fairly similar to whatever other code was using the Reflections class:

`Reflections reflections = new VersionCompliantReflections(new ConfigurationBuilder()
    .setScanners(new SubTypesScanner(false), new ResourcesScanner())
    .setUrls(ClasspathHelper.forClassLoader(classLoadersList.toArray(new ClassLoader[0])))
    .filterInputsBy(filterBuilder));
//do whatever else you were going to do with your Reflections object`

**Comment by ronmamo** [comment]
guava dependency removed.
fixed in 0.9.12
